### PR TITLE
Faster LSTM

### PR DESCRIPTION
**Pull request by carlthome**
On the imdb_bidirectional_lstm.py example the result is a **40% speedup.** from around 57 seconds down to 36 seconds per epoch on a GTX 980 Ti with the Theano backend.

@NasenSpray's [comments about slow LSTM performance](https://github.com/fchollet/keras/issues/2088#issuecomment-201765373) made me curious so I looked into the LSTM implementation and found that eight small GEMMs were performed when really only two large matrix multiplications should be needed because everything is independent before applying the non-linearities. Changing this would require a lot less communication with the GPU at a given timestep.

image
[(source)](http://deeplearning.net/tutorial/lstm.html)

I've implemented this under `consume_less='derp'` at the moment, so it can be compared to the existing modes `consume_less='cpu'` and `consume_less='mem'`.

As I'm usually a bit naive and sloppy though, naturally I worry the math is off! Especially @yaringal's dropout which I've only glanced at. Fortunately the LSTM examples have similar accuracy regardless of `consume_less` setting. Could someone help me verify this a bit more? I'd like someone with the TensorFlow backend to try this out too. I'm on Theano.

### CODE REVIEW

**Comment by fchollet** [comment]
Currently I cannot run it either with Theano or TensorFlow (getting two different error messages, both originating in `K.rnn`). But it sounds interesting.

Did you test it on GPU?

**Comment by carlthome** [comment]
Yes, GPU with Theano. I haven't tried it on CPU.

Could you paste the stack traces please?

**Comment by fchollet** [comment]
Then maybe test it on CPU first.

Theano:
```
  File "/Library/Python/2.7/site-packages/Keras-1.0.1-py2.7.egg/keras/backend/theano_backend.py", line 670, in rnn
    go_backwards=go_backwards)
  File "/Library/Python/2.7/site-packages/theano/scan_module/scan.py", line 1041, in scan
    scan_outs = local_op(*scan_inputs)
  File "/Library/Python/2.7/site-packages/theano/gof/op.py", line 611, in __call__
    node = self.make_node(*inputs, **kwargs)
  File "/Library/Python/2.7/site-packages/theano/scan_module/scan_op.py", line 544, in make_node
    inner_sitsot_out.type.ndim))
ValueError: When compiling the inner function of scan (the function called by scan in each of its iterations) the following error has been encountered: The initial state (`outputs_info` in scan nomenclature) of variable IncSubtensor{Set;:int64:}.0 (argument number 1) has 3 dimension(s), while the corresponding variable in the result of the inner function of scan (`fn`) has 3 dimension(s) (it should be one less than the initial state). For example, if the inner function of scan returns a vector of size d and scan uses the values of the previous time-step, then the initial state in scan should be a matrix of shape (1, d). The first dimension of this matrix corresponds to the number of previous time-steps that scan uses in each of its iterations. In order to solve this issue if the two varialbe currently have the same dimensionality, you can increase the dimensionality of the variable in the initial state of scan by using dimshuffle or shape_padleft.
```

TensorFlow:
```
  File "/Library/Python/2.7/site-packages/Keras-1.0.1-py2.7.egg/keras/backend/tensorflow_backend.py", line 766, in rnn
    output, states = step_function(input, states + constants)
  File "/Library/Python/2.7/site-packages/Keras-1.0.1-py2.7.egg/keras/layers/recurrent.py", line 780, in step
    z = K.dot(x * B_W, self.W) + K.dot(h_tm1 * B_U, self.U) + self.b
  File "/Library/Python/2.7/site-packages/tensorflow/python/ops/math_ops.py", line 467, in binary_op_wrapper
    y = ops.convert_to_tensor(y, dtype=x.dtype.base_dtype, name="y")
  File "/Library/Python/2.7/site-packages/tensorflow/python/framework/ops.py", line 529, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/Library/Python/2.7/site-packages/tensorflow/python/ops/constant_op.py", line 178, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/Library/Python/2.7/site-packages/tensorflow/python/ops/constant_op.py", line 161, in constant
    tensor_util.make_tensor_proto(value, dtype=dtype, shape=shape))
  File "/Library/Python/2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 319, in make_tensor_proto
    _AssertCompatible(values, dtype)
  File "/Library/Python/2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 259, in _AssertCompatible
    (dtype.name, repr(mismatch), type(mismatch).__name__))
TypeError: Expected float32, got list containing Tensors of type '_Message' instead.
```

**Comment by fchollet** [comment]
As a matter of comparison, what are the numbers that you are observing with Theano on GPU, for modes `cpu`, `mem`, `derp`? On what GPU? What about CPU?

**Comment by carlthome** [comment]
I'll do some more thorough benchmarks with the three LSTM examples and look into possible CPU problems. I'm just waiting on CI to spank my ass first.

**Comment by carlthome** [comment]
Surprisingly enough I had no issue running derp mode on CPU with Theano. Also, Travis is green including the [LSTM test](https://github.com/fchollet/keras/blob/master/tests/keras/layers/test_recurrent.py#L111). I wonder how our systems differ. Are you using the master branch of Theano from their git repo or the PyPI build? Could you share the exact code that causes the errors so I can try to reproduce it?

**Comment by fchollet** [comment]
I'm using `0.8.1`, PyPI release.

**Comment by carlthome** [comment]
Interesting. Both Travis and my system use the latest master branch of Theano. I'll try with the PyPI release now to see if that causes problems.

By the way, the speedup on CPU is significant too (around 15%).

**Average run times per epoch on `imdb_bidirectional_lstm.py`:**
- derp 57.5
- cpu 67.5
- mem 67.75

**Comment by fchollet** [comment]
I get the same error with the Theano master branch. I suppose a difference would be the OS; I'm on OSX.

**Comment by carlthome** [comment]
Hmm... OS shouldn't matter? Are you using float64 or float32?

For reference, working .theanorc:

```
[global]
floatX = float32
device = cpu
mode = FAST_RUN

[compile]
mode = FAST_RUN

[lib]
cnmem = 0.8

[scan]
allow_gc = False
```

Perhaps gc is the culprit? I'm clueless.

**Comment by carlthome** [comment]
I've found the problem. Fix soon.

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -691,6 +690,7 @@ def build(self, input_shape):
                                        name='{}_U_o'.format(self.name))
             self.b_o = K.zeros((self.output_dim,), name='{}_b_o'.format(self.name))
 
+
```
Comment (carlthome) [inline keras/layers/recurrent.py]: PEP! :cactus:

**Comment by carlthome** [comment]
Alright, fixed. Try it now please.

The dropout constants are passed to `step()` as normal Python lists and not tensors, which I did not expect.

As it stands currently a limitation of doing two large GEMMs is that the same dropout will be used for all gates but this could be remedied.

[The forget gate initialization is also wrong at the moment](https://github.com/carlthome/keras/blob/master/keras/layers/recurrent.py#L653) as its all zeros. It would be an easy fix if the backend supported subtensor assignment as suggested by https://github.com/fchollet/keras/pull/2093.

By the way, there is no unit test for RNN dropout.

**Comment by carlthome** [comment]
All rrrors are related to this https://github.com/fchollet/keras/commit/c9f7d970e97d5af3f92a96cac75554aaf4261337

**Comment by joelthchao** [comment]
I think these errors are resulted by parallel issue. I may spend some time revising a better `test_image.py`. You probably can submit it again and hope the error won't happen again.

**Comment by yaringal** [comment]
> As it stands a limitation of doing two large GEMMs is that the same dropout will be used for all gates but this could probably be remedied. Ideas welcome.
yep - I spotted that as well. I played with this approach a bit some time ago, and I think it leads to slightly worse performance. Is there a way to do the large GEMM with different masks for the 4 gates? A naive approach would be to zero out parts of the rows of the large matrix, but there might be a more efficient way of doing this.

**Comment by carlthome** [comment]
@joelthchao cheers! Worked fine. It seems insane to me that CI could fail seemingly at random though.

**Comment by carlthome** [comment]
@yaringal what about applying the masks directly to the weight matrices instead? Is that sensible? Something like: `z = K.dot(x, self.W * B_W) + K.dot(h_tm1, self.U * B_U) + self.b`

**Comment by yaringal** [comment]
That would result in the same thing... it will also increase the variance of the estimator (when applying the mask to x we have different Bernoulli samples for each dimensions and each point in the batch; when applying the mask to W all batch points will have to use the same mask).

**Comment by carlthome** [comment]
@yaringal, would it really result in the same result though? The B_W mask would have four times as many coin flips than B_W[0] so at least each gate's weights would be masked differently, right? Or maybe I'm just misunderstanding the matrix multiplication dimensions. Anyway, regarding having all batch points getting the same mask though, yeah...

Still, a GPU performance increase of roughly 40% at the expense of requiring the same RNN dropout for all gates is useful in many cases, especially considering many don't even use dropout, and considering some might be looking to [RNN batch normalization](http://arxiv.org/abs/1510.01378) in the future. I propose merging a third `consume_less` mode (renamed to something better than derp of course, maybe `consume_less='time'`? :joy: ). Memory usage remains the same, and there's even a slight CPU performance increase (roughly 15%). This is all in line with Theano's manual.

I'd like the forget_bias_init to be set properly first though. Preferably with this https://github.com/fchollet/keras/pull/2093 but optionally be allowing K.concatenate() with a name.

**Comment by yaringal** [comment]
@carlthome yeah, that's a sensible trade-off. I'm re-running experiments with the same RNN dropout for all gates to quantify the accuracy trade-off. I'm gonna play with some other ideas I have to make this more efficient.

**Comment by fchollet** [comment]
> I propose merging a third consume_less mode (renamed to something better than derp of course, maybe consume_less='time'

Why not make it part of `consume_less='cpu'`? What would be the downsides?

**Comment by fchollet** [comment]
> I'd like the forget_bias_init to be set properly first though. 

This is easy, tensor initialization is done via Numpy arrays of initial values.

**Comment by fchollet** [comment]
Also, if there a significant speed up on GPU but not on CPU, `consume_less='gpu'` would be a user-friendly mode name. By the way, it would be great to test it on more networks / tasks to see how the speed-up fares across a range of problems.

**Comment by carlthome** [comment]
> Why not make it part of consume_less='cpu'? What would be the downsides?

@fchollet, the downside is that you couldn't use the exactly correct RNN dropout, but rather might get worsened regularization (and possibly other side effects) because all gates need to share the same dropout, because of the GEMM concatenation.

> This is easy, tensor initialization is done via Numpy arrays of initial values.

Alright, I'll readd the forget_bias_init tomorrow just like that.

> Also, if there a significant speed up on GPU but not on CPU, consume_less='gpu' would be a user-friendly mode name. By the way, it would be great to test it on more networks / tasks to see how the speed-up fares across a range of problems.

`consume_less='gpu'` sounds good. I think there might be some value to keeping the previous two modes, because of the dropout, but I guess it really should come down to what @yaringal quantifies in his experiments. Maybe it doesn't make much of a real-world difference?

**Comment by yaringal** [comment]
@carlthome I ran a LM model experiment using Zaremba et al. (2014)'s large model setting (Pen Treebank with 2 layer LSTM, 1250 units in each layer) with the two dropout variants: 1) current implementation (each gate using different masks), 2) tied masks (using a large matrix product and a single dropout mask for each input - `K.dot(x * B_W[0], self.W)` with W=[W1; W2; W3; W4]). I repeated each experiment 3 times with different random seeds. 
I got: 
* 79.14+-0.18 validation perplexity and 76.28+-0.33 test perplexity for setting 1
* 79.29+-0.07 validation perplexity and 76.66+-0.08 test perplexity for setting 2
* (this is in comparison to Zaremba et al.'s 82.2 validation perplexity and 78.4 test perplexity)

I'd say the difference is statistically significant but negligible, so either one can be used. 
Btw, if you're interested in the theoretical interpretation of setting 2, it simply corresponds to a different approximating distribution structure. I'll add these results to the arXiv paper.

**Comment by fchollet** [comment]
@carlthome I think we should move forward with this and merge it as a new mode, `consume_less='gpu'`. It would also be great to have CPU/GPU benchmarks on a couple tasks (something we could add to the documentation to help users make a choice).

**Comment by carlthome** [comment]
Great! Let me do some code cleanup and I'll notify you for a code review + merge in a couple of hours.

**Comment by xingdi-eric-yuan** [comment]
Does something similar also work for GRU?

**Comment by carlthome** [comment]
@xingdi-eric-yuan Yes.

**Comment by carlthome** [comment]
Added an example showcasing the difference between `consume_less` modes.
image
One could certainly wonder why I managed to pick hyperparameters that lead to increasing validation loss, but I was more concerned with showcasing computational performance.  :dancer:  

I thought about including RAM and CPU load with psutils but profiling is a bit of a hassle (multi-threading, at the very least) and I don't know if it's smart to introduce new external dependencies just for an example. Also, profiling GPUs is almost impossible since there is no clear hardware-agnostic solution out there. Best to leave this up to each user.

Finally, I think it's a little weird to call it consume_less='gpu' when it's really doing quite the opposite and actually is managing to consume more of the GPU resources (which is a good thing), but I don't have any other suggestions.

**Comment by carlthome** [comment]
@fchollet green on Travis. Please code review.

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -743,21 +758,36 @@ def step(self, x, states):
...
+                x_f = x[:, self.output_dim: 2 * self.output_dim]
+                x_c = x[:, 2 * self.output_dim: 3 * self.output_dim]
+                x_o = x[:, 3 * self.output_dim:]
+            elif self.consume_less == 'mem':
+                x_i = K.dot(x * B_W[0], self.W_i) + self.b_i
+                x_f = K.dot(x * B_W[1], self.W_f) + self.b_f
+                x_c = K.dot(x * B_W[2], self.W_c) + self.b_c
+                x_o = K.dot(x * B_W[3], self.W_o) + self.b_o
+            else:
+                raise Exception('Unknown `consume_less` mode.')
```
Comment (wxs) [inline keras/layers/recurrent.py]: There's no reason for these three branches to be inside the `else` of line 773, rather that else should be removed should all be `elif`s up one layer, i.e.

```
if self.consume_less == 'gpu':
    ...
elif self.consume_less == 'cpu':
    ...
elif self.consume_less == 'mem':
    ...
else:
    raise Exception('Unknown `consume_less` mode.')
```

Comment (xingdi-eric-yuan) [inline keras/layers/recurrent.py]: cause he's sharing line 787-790 for 'cpu' and 'mem'

**Comment by fchollet** [comment]
> Does something similar also work for GRU?

We should make this mode available for GRU as well, and make clear in the docs that it's available for LSTM and GRU but not for SimpleRNN.

> Finally, I think it's a little weird to call it consume_less='gpu' when it's really quite the opposite and actually is managing to consume more GPU resources (which is a good thing), but I don't have any other suggestions.

Details aside, all we care about is to clearly convey to users when they should be using this mode. I think "consume_less='gpu'" is a decent name in that regard (you should interpret it as "consume less GPU time", not as "lower GPU utilization").

Thanks for the graphs, this is very informative. I assume this is on GPU; do you have the same for CPU? Also, what are these hyperparameters you mention? I would suppose that batch_size and model size could have a significant impact on these profiles btw.

File: keras/layers/recurrent.py (hunk 2 above)
Comment (wxs) [inline keras/layers/recurrent.py]: oops yeah I can't read apparently. Disregard!

**Comment by xingdi-eric-yuan** [comment]
Did similar thing on GRU, the performance does not improve as much as on LSTM, may because the "hh" part.
Using gpu device 1: GeForce GTX TITAN X (CNMeM is disabled, cuDNN 4007)
Loading data...
Testing mode: consume_less='cpu'
Train on 20000 samples, validate on 5000 samples
Epoch 1/10
18s - loss: 0.6862 - acc: 0.6068 - val_loss: 0.4884 - val_acc: 0.7606
Epoch 2/10
18s - loss: 0.4525 - acc: 0.7907 - val_loss: 0.4510 - val_acc: 0.7986
Epoch 3/10
18s - loss: 0.3402 - acc: 0.8554 - val_loss: 0.4445 - val_acc: 0.8204
Epoch 4/10
19s - loss: 0.2641 - acc: 0.8937 - val_loss: 0.4725 - val_acc: 0.8192
Epoch 5/10
19s - loss: 0.2024 - acc: 0.9216 - val_loss: 0.5529 - val_acc: 0.8154
Epoch 6/10
19s - loss: 0.1599 - acc: 0.9394 - val_loss: 0.6228 - val_acc: 0.8192
Epoch 7/10
20s - loss: 0.1370 - acc: 0.9480 - val_loss: 0.6132 - val_acc: 0.8224
Epoch 8/10
19s - loss: 0.1212 - acc: 0.9545 - val_loss: 0.6647 - val_acc: 0.8166
Epoch 9/10
19s - loss: 0.1106 - acc: 0.9580 - val_loss: 0.6520 - val_acc: 0.8164
Epoch 10/10
19s - loss: 0.0841 - acc: 0.9698 - val_loss: 0.7566 - val_acc: 0.8172
Testing mode: consume_less='mem'
Train on 20000 samples, validate on 5000 samples
Epoch 1/10
19s - loss: 0.6526 - acc: 0.6249 - val_loss: 0.4794 - val_acc: 0.7712
Epoch 2/10
19s - loss: 0.4449 - acc: 0.7948 - val_loss: 0.4351 - val_acc: 0.8020
Epoch 3/10
19s - loss: 0.3393 - acc: 0.8565 - val_loss: 0.4940 - val_acc: 0.8048
Epoch 4/10
18s - loss: 0.2585 - acc: 0.8956 - val_loss: 0.4660 - val_acc: 0.8230
Epoch 5/10
19s - loss: 0.1982 - acc: 0.9224 - val_loss: 0.4876 - val_acc: 0.8184
Epoch 6/10
18s - loss: 0.1643 - acc: 0.9371 - val_loss: 0.5996 - val_acc: 0.8100
Epoch 7/10
18s - loss: 0.1362 - acc: 0.9493 - val_loss: 0.5533 - val_acc: 0.8212
Epoch 8/10
18s - loss: 0.1093 - acc: 0.9592 - val_loss: 0.6466 - val_acc: 0.8178
Epoch 9/10
18s - loss: 0.1031 - acc: 0.9624 - val_loss: 0.6837 - val_acc: 0.8140
Epoch 10/10
18s - loss: 0.0972 - acc: 0.9652 - val_loss: 0.7122 - val_acc: 0.8152
Testing mode: consume_less='gpu'
Train on 20000 samples, validate on 5000 samples
Epoch 1/10
17s - loss: 0.6306 - acc: 0.6531 - val_loss: 0.4628 - val_acc: 0.7870
Epoch 2/10
17s - loss: 0.4264 - acc: 0.8069 - val_loss: 0.4330 - val_acc: 0.8044
Epoch 3/10
17s - loss: 0.3364 - acc: 0.8572 - val_loss: 0.4474 - val_acc: 0.8152
Epoch 4/10
17s - loss: 0.2416 - acc: 0.9030 - val_loss: 0.4387 - val_acc: 0.8196
Epoch 5/10
17s - loss: 0.1909 - acc: 0.9247 - val_loss: 0.5168 - val_acc: 0.8086
Epoch 6/10
17s - loss: 0.1521 - acc: 0.9414 - val_loss: 0.5398 - val_acc: 0.8194
Epoch 7/10
17s - loss: 0.1350 - acc: 0.9487 - val_loss: 0.6338 - val_acc: 0.8188
Epoch 8/10
17s - loss: 0.1068 - acc: 0.9603 - val_loss: 0.6164 - val_acc: 0.8206
Epoch 9/10
17s - loss: 0.0973 - acc: 0.9651 - val_loss: 0.7144 - val_acc: 0.8204
Epoch 10/10
17s - loss: 0.0912 - acc: 0.9674 - val_loss: 0.6962 - val_acc: 0.8188

**Comment by NasenSpray** [comment]
@xingdi-eric-yuan: could you retry with CNMeM (and allow_gc) enabled?

**Comment by xingdi-eric-yuan** [comment]
@NasenSpray 
both with cnmem enabled (lib.cnmem=0.4, sharing memory with other process, but the other process is suspended):
average time when allow_gc=**True**: 
cpu/mem/gpu
15s/16s/14s
average time when allow_gc=**False**:
cpu/mem/gpu
14.7s/16s/14s

**Comment by carlthome** [comment]
>We should make this mode available for GRU as well, and make clear in the docs that it's available for LSTM and GRU but not for SimpleRNN.

Anyone up for adding this? @xingdi-eric-yuan?

>Details aside, all we care about is to clearly convey to users when they should be using this mode. I think "consume_less='gpu'" is a decent name in that regard (you should interpret it as "consume less GPU time", not as "lower GPU utilization").

So, _consume less memory time_? I would have called it `optimize_for='cpu'|'mem'|'gpu'` but breaking the API is bad.

>Thanks for the graphs, this is very informative. I assume this is on GPU; do you have the same for CPU?

Cheers! Yes, GPU. On CPU (quad-core with hyper-threading):

image

Feel free to run it on your machine: https://github.com/carlthome/keras/blob/master/examples/lstm_benchmark.py

>Also, what are these hyperparameters you mention?

- I mostly scaled up the embedding dimensionality to make the problem more demanding such that each epoch would take long enough so that elapsed time could be measured reliably with Python's own `time`.
- Particularly batch normalization and high dropout in both the embedding and the LSTM seem to have made the model perform a bit bonkers on validation data (oscillating loss, despite fitting training data as usual) after a couple of epochs, but I figured it was neat to have examples that show off some of the cool builtin layers and features.

>I would suppose that batch_size and model size could have a significant impact on these profiles btw.

I think model size should scale pretty linearly in computational time and memory usage for the three modes. Hard to test though. Batch size seems important. Intuitively the batch size would have to be large enough for the GEMMs to become the bottleneck and not the transferring of data between RAM and VRAM.

**Comment by xingdi-eric-yuan** [comment]
> We should make this mode available for GRU as well, and make clear in the docs that it's available for LSTM and GRU but not for SimpleRNN.
Anyone up for adding this? @xingdi-eric-yuan?

@carlthome Sure, how should I do that? a new pull request? or we merge the gru part into this PR?

**Comment by carlthome** [comment]
Either way is fine by me. I could grant you access to my repo and you'd add it to this pr (@fchollet's prefered way maybe?), or you just do a new one in your own repo (preferably after this has been merged just to avoid manual merge conflicts). :tada:

**Comment by xingdi-eric-yuan** [comment]
ok then I'll do a new PR after this is merged,  by then more people can check my code 👍

**Comment by fchollet** [comment]
> I would have called it optimize_for='cpu'|'mem'|'gpu' but breaking the API is bad.

Yes, in retrospect that may have been a better name. But breaking the API for cosmetic reasons is not something we do.

> ok then I'll do a new PR after this is merged, by then more people can check my code

Ok, cool.

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -637,65 +645,72 @@ def __init__(self, output_dim,
...
-                                   name='{}_U_o'.format(self.name))
-        self.b_o = K.zeros((self.output_dim,), name='{}_b_o'.format(self.name))
+        if self.consume_less == 'gpu':
+            self.W = self.init((self.input_dim, 4*self.output_dim),
+                               name='{}_W'.format(self.name))
+            self.U = self.inner_init((self.output_dim, 4*self.output_dim),
+                                     name='{}_U'.format(self.name))
+
+            self.b = K.variable(np.hstack((np.zeros(self.output_dim),
+                                           self.forget_bias_init(self.output_dim).get_value(),
```
Comment (fchollet) [inline keras/layers/recurrent.py]: This is Theano syntax and breaks with TF. Use `K` instead.

**Comment by fchollet** [comment]
BTW there are no tests for this feature, and in fact it would fail them if there were any. Please loop through the different implementation modes in the RNN test.

File: keras/layers/recurrent.py [hunk 4]
```diff
@@ -660,7 +660,7 @@ def build(self, input_shape):
                                      name='{}_U'.format(self.name))
 
             self.b = K.variable(np.hstack((np.zeros(self.output_dim),
-                                           self.forget_bias_init(self.output_dim).get_value(),
+                                           K.get_value(self.forget_bias_init(self.output_dim)),
```
Comment (carlthome) [inline keras/layers/recurrent.py]: Duck typing...

**Comment by carlthome** [comment]
@fchollet merge?

**Comment by fchollet** [comment]
LGTM

**Comment by xingdi-eric-yuan** [comment]
my PR is here: https://github.com/fchollet/keras/pull/2633

**Comment by faroit** [comment]
I really enjoy the speedup, thanks! 

But since I try to have my experiments flexible so that I can debug on CPU and run on GPU the ```consume_less``` argument adds additional complexity to the development.

What about a fourth mode which would be ```auto``` and detects if keras runs on gpu or cpu and chooses the best setting automatically. I could provide a PR if this would be helpful.

**Comment by carlthome** [comment]
@faroit, considering `consume_less='gpu'` is consistently faster, even when running on a CPU, I'd just leave the setting at that.

Long term I hope to see most of [these optimizations](https://devblogs.nvidia.com/parallelforall/optimizing-recurrent-neural-networks-cudnn-5/) worked into Keras and for the modes to be merged into a single highly-optimized implementation. Much of this must be done in low-level code though and is not a task for Keras but rather Theano and TensorFlow.

**Pull request merged**